server/container_create.go

```diff
@@ -487,6 +479,50 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
 		specgen.SetProcessCwd(cwd)
 	}
 
+	// Setup user and groups
+	if linux != nil {
+		if linux.GetSecurityContext() != nil {
+			containerUser := ""
+			// Case 1: run as user is set by kubelet
+			if linux.GetSecurityContext().RunAsUser != nil {
+				containerUser = strconv.FormatInt(linux.GetSecurityContext().GetRunAsUser().Value, 10)
+			} else {
+				// Case 2: run as username is set by kubelet
+				userName := linux.GetSecurityContext().RunAsUsername
+				if userName != "" {
+					containerUser = userName
+				} else {
+					// Case 3: get user from image config
+					imageUser := containerInfo.Config.Config.User
+					if imageUser != "" {
+						containerUser = imageUser
+					}
+				}
+			}
+
+			logrus.Debugf("CONTAINER USER: %+v", containerUser)
+
+			// Add uid, gid and groups from user
+			uid, gid, addGroups, err1 := getUserInfo(mountPoint, containerUser)
+			if err != nil {
+				return nil, err1
+			}
+
+			logrus.Debugf("UID: %v, GID: %v, Groups: %+v", uid, gid, addGroups)
+			specgen.SetProcessUID(uid)
+			specgen.SetProcessGID(gid)
+			for _, group := range addGroups {
+				specgen.AddProcessAdditionalGid(group)
+			}
+
+			// Add groups from CRI
+			groups := linux.GetSecurityContext().SupplementalGroups
+			for _, group := range groups {
+				specgen.AddProcessAdditionalGid(uint32(group))
+			}
+		}
+	}
+
```

Comment:
The code looks good to me, but given the `createSandboxContainer()` current length it would make sense to wrap that implementation into e.g. a `setUserGroup()` separate function.